docs/developer/decision-records/2024-10-09-vault-authentication-refactor/README.md

```diff
@@ -0,0 +1,194 @@
+# Hashicorp Vault authentication refactor
+
+## Decision
+
+Refactor the HashiCorp Vault extension and make it extensible with different methods of authentication.
+
+## Rationale
+
+The current implementation of the authentication in the HashiCorp Vault extension has no way to add new authentication methods and to switch between them.
+The full list of possible authentication methods can be found in the [HashiCorp Vault Documentation](https://developer.hashicorp.com/vault/docs/auth)
+Relevant examples are:
+
+* [Token auth](https://developer.hashicorp.com/vault/docs/auth/token)
+* [Kubernetes auth](https://developer.hashicorp.com/vault/docs/auth/kubernetes)
+* [AppRole auth](https://developer.hashicorp.com/vault/docs/auth/approle)
+It goes against the EDC extensibility model and needs to be remedied.
+Additionally, extracting the current implementation of the token authentication into its own extension will improve readability and maintainability of the code.
+
+## Approach
+The refactor will affect only the `vault-hashicorp` extension.
+
+To allow an extensible authentication module for HashiCorp Vault, the implementation will follow the registry pattern.
+
+### Vault Auth SPI
+
+To allow the proper organisation of the interfaces needed for the refactoring, a new module named `vault-auth-spi` is introduced in the `spi` directory.
+
+
+### Vault Authentication Auth interface
+
+To interact with a multitude of different authentication methods, an interface for the Vault Authentication Service is needed.
+This interface will contain two methods:
+
+* `login()`
+* `getVaultToken()`
+
+
+
+```java
+public interface HashicorpVaultAuth {
+
+    // Contains the authentication and generates the `client_token` from HashiCorp Vault in accordance to the chosen auth method.
+    void login();
+    
+    // The stored token is returned.
+    String vaultToken();
+    
+}
+```
+
+This interface will later be registered in a registry and used by the `HashicorpVaultClient` to receive the `client_token` for the request authentication.
```

Comment:
an interface cannot be registered, you mean interface implementations right? Maybe this explanation can be made in the "registry" section directly/